File: scala/kindergarten-garden/src/main/scala/Garden.scala
object Plant extends Enumeration {
  type Plant = Value
  val Radishes, Clover, Grass, Violets = Value
}

class Garden(val students: Seq[String], val sillString:String) {
  import Plant._
  import Garden._

  lazy val indexByStudent =
    students.sorted
      .zipWithIndex
      .groupBy(_._1)
        .mapValues(_.head._2 + 1)

  lazy val sills = sillString.split("\n").toList

  def plantsByIndex(i: Int, width: Int): Seq[Plant] =
    sills
      .map(_.slice(i-width, i))
      .flatten
      .map(Plants)

  def getPlants(student: String): Seq[Plant] =
    indexByStudent.get(student)
      .map(_*2)
      .map(plantsByIndex(_:Int, 2))
      .getOrElse(Nil)
}

object Garden {
  val DefaultStudents =
    Seq("Alice", "Bob", "Charlie", "David", "Eve", "Fred", "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry")

  val Plants =
    Plant.values
      .groupBy(_.toString.charAt(0))
      .mapValues(_.head)

  def defaultGarden(sill: String) = Garden(DefaultStudents, sill)

  def apply(students: Seq[String], sill: String) = new Garden(students, sill)

}

File: haskell/kindergarten-garden/src/Garden.hs
module Garden
    ( Plant (..)
    , defaultGarden
    , garden
    , lookupPlants
    ) where

import qualified Data.List as L
import Data.List.Split (chunksOf)
import Data.Function (on)
import qualified Data.Map as M

data Plant = Clover
           | Grass
           | Radishes
           | Violets
           deriving (Eq, Show)

toPlant :: Char -> Plant
toPlant x
  | x == 'C' = Clover
  | x == 'G' = Grass
  | x == 'R' = Radishes
  | x == 'V' = Violets
  | otherwise = error $ x : ": unknown plant"

toPlants :: [Char] -> [Plant]
toPlants = map toPlant

defaultStudents :: [String]
defaultStudents =
    ["Alice", "Bob", "Charlie", "David",
     "Eve", "Fred", "Ginny", "Harriet",
     "Ileana", "Joseph", "Kincaid", "Larry"]

defaultGarden :: String -> M.Map String [Plant]
defaultGarden = garden defaultStudents

garden :: [String] -> String -> M.Map String [Plant]
garden xs = M.fromList . zip (L.sort xs) . parseGarden
  where
    combine (ys, zs) = ys ++ zs
    splitOnNl = fmap (drop 1) . span (/='\n')
    splitAndZipChunks = uncurry (zip `on` chunksOf 2) . splitOnNl
    parseGarden = map (toPlants . combine) . splitAndZipChunks

lookupPlants :: String -> M.Map String [Plant] -> [Plant]
lookupPlants = flip (M.!)
